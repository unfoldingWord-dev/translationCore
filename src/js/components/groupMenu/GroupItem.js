import React from 'react';
import ReactTooltip from 'react-tooltip';
import PropTypes from 'prop-types';
import style from './Style';

class GroupItem extends React.Component {
  /**
   * @description Generate the proper glyphicon based on selections
   * @return {component} statusGlyph - component to render
   */
  constructor(props) {
    super(props);
    this.onClick = this.onClick.bind(this);
  }

  componentDidMount() {
    if (this.props.active) {
      if (this.props.inView(this.props.groupMenuHeader, this)) {
        //If the menu and current check are able to be rendered in the
        //same window scroll to the group menu item
        this.props.scrollIntoView(this.props.groupMenuHeader);
      }
      else {
        //Scroll to the current check item
        this.props.scrollIntoView(this);
      }
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.active) {
      if (this.props.inView(nextProps.groupMenuHeader, this)) {
        //If the menu and current check are able to be rendered in the
        //same window scroll to the group menu item
        nextProps.scrollIntoView(nextProps.groupMenuHeader);
      }
      else {
        //Scroll to the current check item
        nextProps.scrollIntoView(this);
      }
    }
  }

  onClick() {
    this.props.actions.changeCurrentContextId(this.props.contextId);
  }

  render() {
    const { reference } = this.props.contextId;
    return (
      <div onClick={this.onClick}
        className={"group-item"+(this.props.active?" active":"")}
        style={this.props.active ? style.activeSubMenuItem : style.subMenuItem}>
        {this.props.statusBadge}
        <span
          className="selection"
          data-tip={this.props.selectionText}
          data-place="bottom"
          data-effect="float"
          data-type="dark"
          data-class="selection-tooltip"
          data-delay-hide="100">
          {reference.chapterVerseMenu ?
            <span style={style.groupItemText}>
              {`${reference.text} ${reference.verse}`}
            </span>
            :
            <span style={style.groupItemText}>
              {" " + this.props.bookName + " " + reference.chapter + ":" + reference.verse + " " + this.props.selectionText}
            </span>
          }
        </span>
        <ReactTooltip />
      </div>
    );
  }
}

GroupItem.propTypes = {
    bookName: PropTypes.string.isRequired,
    selectionText: PropTypes.string.isRequired,
    contextId: PropTypes.object.isRequired,
    actions: PropTypes.shape({
        changeCurrentContextId: PropTypes.func.isRequired
    }),
    statusBadge: PropTypes.object.isRequired,
    scrollIntoView: PropTypes.func.isRequired,
    inView: PropTypes.func.isRequired,
    active: PropTypes.bool.isRequired,
    groupMenuHeader: PropTypes.object
};

module.exports = GroupItem;
