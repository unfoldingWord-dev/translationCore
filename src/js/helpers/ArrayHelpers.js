/**
 * @description - Groups consecutive numbers in an array
 * @param {Array} numbers - array of numbers to be grouped
 * @returns {Array} - grouped array of array of consecutive numbers
 */
export const groupConsecutiveNumbers = (numbers) => (
  numbers.reduce(function(accumulator, currentValue, currentIndex, originalArray) {
    if (currentValue) { // ignore undefined entries
      // if this iteration is consecutive to the last, add it to the previous run
      if (currentValue - originalArray[currentIndex - 1] === 1) {
        accumulator[accumulator.length - 1].push(currentValue);
      } else { // the start of a new run including first element
        // create a new subarray with this as the start
        accumulator.push([currentValue]);
      }
    }
    return accumulator; // return state for next iteration
  }, [])
);
/**
 * @description - Deletes indices from an array safely
 * @param {Array} array - array elements to delete from
 * @returns {Array} - the resulting array after indexes were safely removed
 */
export const deleteIndices = (array, indices) => {
  let _array = JSON.parse(JSON.stringify(array));
  indices.sort( (a,b) => b - a );
  indices.forEach(index => {
    if (index >= 0) _array.splice(index, 1);
  });
  return _array;
};
