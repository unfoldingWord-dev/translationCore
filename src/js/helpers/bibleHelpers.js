/* eslint-disable no-console */
import Path from 'path-extra';
import fs from 'fs-extra';
import BooksOfBible from '../../../tcResources/books';
import {getLatestVersionInPath} from "./ResourcesHelpers";

/**
 *
 * @param {string} bookAbbr - The book abbreviation to convert
 */
export function convertToFullBookName(bookAbbr) {
  if (!bookAbbr) return;
  return BooksOfBible[bookAbbr.toString().toLowerCase()];
}

/**
 *
 * @param {string} projectBook - Book abbreviation
 */
export function isOldTestament(projectBook) {
  var passedBook = false;
  for (var book in BooksOfBible) {
    if (book === projectBook) passedBook = true;
    if (BooksOfBible[book] === 'Malachi' && passedBook) {
      return true;
    }
  }
  return false;
}

/**
 * determine
 * @param bookId
 * @return {{resourceLanguage: string, bibleID: string}}
 */
export function getOLforBook(bookId) {
  const isOT = isOldTestament(bookId);
  const languageId = (isOT) ? 'he' : 'grc';
  const bibleId = (isOT) ? 'uhb' : 'ugnt';
  return {languageId, bibleId};
}

/**
 * This checks if a project is missing verses.
 * @param projectDir
 * @param bookId
 * @param resourceDir
 * @return {boolean}
 */
export const isProjectMissingVerses = (projectDir, bookId, resourceDir) => {
  try {
    let languageId = 'en';
    const resourcePath = Path.join(resourceDir, languageId, 'bibles', 'ult');
    const versionPath = getLatestVersionInPath(resourcePath) || resourcePath;
    const indexLocation = Path.join(versionPath, 'index.json');
    const expectedVerses = fs.readJSONSync(indexLocation);
    const actualVersesObject = {};
    const currentFolderChapters = fs.readdirSync(Path.join(projectDir, bookId));
    let chapterLength = 0;
    for (let currentChapterFile of currentFolderChapters) {
      let currentChapter = Path.parse(currentChapterFile).name;
      if (!parseInt(currentChapter)) continue;
      chapterLength++;
      let verseLength = 0;
      try {
        let currentChapterObject = fs.readJSONSync(
          Path.join(projectDir, bookId, currentChapterFile)
        );
        for (let verseIndex in currentChapterObject) {
          let verse = currentChapterObject[verseIndex];
          if (verse && verseIndex > 0) verseLength++;
        }
      } catch (e) {
        console.warn(e);
      }
      actualVersesObject[currentChapter] = verseLength;
    }
    actualVersesObject.chapters = chapterLength;
    let currentExpectedVerese = expectedVerses[bookId];
    return (
      JSON.stringify(currentExpectedVerese) !==
      JSON.stringify(actualVersesObject)
    );
  } catch (e) {
    console.warn(
      'ult index file not found missing verse detection is invalid. Please delete ~/translationCore/resources folder'
    );
    return false;
  }
};
