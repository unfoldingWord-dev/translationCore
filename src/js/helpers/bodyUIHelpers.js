export function getIconColorFromIndex(stepIndex) {
  let colorsArray = [];
  colorsArray.push(stepIndex === 0 ? "black" : stepIndex > 0 ? "var(--accent-color-dark)" : "");
  colorsArray.push(stepIndex === 1 ? "black" : 1 > stepIndex ? "grey" : stepIndex > 1 ? "var(--accent-color-dark)" : "");
  colorsArray.push(stepIndex === 2 ? "black" : 2 > stepIndex ? "grey" : stepIndex > 2 ? "var(--accent-color-dark)" : "");
  colorsArray.push(stepIndex === 3 ? "black" : 3 > stepIndex ? "grey" : stepIndex > 3 ? "var(--accent-color-dark)" : "");
  return colorsArray;
}