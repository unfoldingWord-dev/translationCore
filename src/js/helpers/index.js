import { generateTimestamp } from './TimestampGenerator.js';
import * as loadHelpers from './LoadHelpers.js';

export {
  generateTimestamp,
  loadHelpers,
};
