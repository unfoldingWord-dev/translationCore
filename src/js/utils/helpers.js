import Path from 'path';
import * as fs from 'fs-extra';
import ManifestGenerator from '../components/core/create_project/ProjectManifest';
import BooksOfBible from '../components/core/BooksOfBible';

module.exports.isOldTestament = function (projectBook) {
    var passedBook = false;
    for (var book in BooksOfBible) {
        if (book == projectBook) passedBook = true;
        if (BooksOfBible[book] == "Malachi" && passedBook) {
            return true;
        }
    }
    return false;
}

module.exports.checkIfValidBetaProject = function (manifest) {
    if (manifest && manifest.project) return manifest.project.id == "eph" || manifest.project.id == "tit";
    else if (manifest && manifest.ts_project) return manifest.ts_project.id == "eph" || manifest.ts_project.id == "tit";
}

/**
 * 
 * @param {*} path 
 * @param {*} file 
 */
export function loadFile(path, file) {
    try {
        var manifest = fs.readJsonSync(Path.join(path, file));
        return manifest;
    }
    catch (e) {
        return null
    }
}

/**
 * @description - Generates and saves a translationCore manifest file
 * @param {string} projectSaveLocation - Filepath of where the translationCore manifest file will
 * be saved. Must be an ABSOLUTE PATH
 * @param {object} data - The translationCore manifest data to be saved
 * @param {object} tsManifest - The translationStudio manifest data loaded from a translation
 * studio project
 */
module.exports.saveManifest = function (projectSaveLocation, link, oldManifest, currentUser) {
    var data = {
        user: [currentUser],
        repo: link
    }
    var manifest;
    try {
        var manifestLocation = Path.join(projectSaveLocation, 'manifest.json');
        if (oldManifest.package_version == '3') {
            //some older versions of ts-manifest have to be tweaked to work
            manifest = this.fixManifestVerThree(oldManifest);
        } else {
            manifest = ManifestGenerator(data, oldManifest);
        }
        fs.outputJsonSync(manifestLocation, manifest);
    }
    catch (err) {
        console.error(err);
    }
    return manifest;
}

/**
 * @desription - Uses the tc-standard format for projects to make package_version 3 compatible
 * @param oldManifest - The name of an employee.
 */
module.exports.fixManifestVerThree = function (oldManifest) {
    var newManifest = {};
    try {
        for (var oldElements in oldManifest) {
            newManifest[oldElements] = oldManifest[oldElements];
        }
        newManifest.finished_chunks = oldManifest.finished_frames;
        newManifest.ts_project = {};
        newManifest.ts_project.id = oldManifest.project_id;
        newManifest.ts_project.name = this.convertToFullBookName(oldManifest.project_id);
        for (var el in oldManifest.source_translations) {
            newManifest.source_translations = oldManifest.source_translations[el];
            var parameters = el.split("-");
            newManifest.source_translations.language_id = parameters[1];
            newManifest.source_translations.resource_id = parameters[2];
            break;
        }
    } catch (e) {
        console.error(e);
    }
    return newManifest;
}

module.exports.convertToFullBookName = function (bookAbbr) {
    if (!bookAbbr) return;
    return BooksOfBible[bookAbbr.toString().toLowerCase()];
}