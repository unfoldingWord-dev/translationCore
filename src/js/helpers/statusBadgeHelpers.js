import React from 'react';
import ReactDOMServer from 'react-dom/server';
import ReactTooltip from 'react-tooltip';
import SVGInline from 'react-svg-inline';
import { Glyphicon } from 'react-bootstrap';
import fs from 'fs-extra';
import path from 'path-extra';
import * as style from '../components/groupMenu/Style';

const SVG_ICONS = ['invalidated'];

export function getGlyphIcons(glyphs) {
  const glyphicons = [];
  if (glyphs && glyphs.length) {
    glyphs.forEach((glyph)=>{
      if (SVG_ICONS.indexOf(glyph) >= 0) {
        const svgPath = path.join(__dirname, '..', '..', 'images', glyph+'.svg');
        const svg = fs.readFileSync(svgPath, 'utf8');
        if (svg)
          glyphicons.push(<SVGInline className={"glyphicon glyphicon-"+glyph} svg={svg} style={style.menuItem.statusIcon.invalidated} />);
      } else {
        glyphicons.push(<Glyphicon key={glyph} glyph={glyph} style={style.menuItem.statusIcon[glyph]} />);
      }
    });
  } else {
    glyphicons.push(<Glyphicon glyph="" style={style.menuItem.statusIcon.blank} />);
  }
  return glyphicons;
}

export function getStatusBadge(glyphs) {
  const statusGlyphs = getGlyphIcons(glyphs);
  const statusCount = statusGlyphs.length;
  const mainGlyph = statusGlyphs.shift();
  if (statusCount > 1) {
    const tooltip = ReactDOMServer.renderToString(statusGlyphs);
    return (
      <div className="status-badge-wrapper">
        <div
          className="status-badge"
          data-tip={tooltip}
          data-html="true"
          data-place="bottom"
          data-effect="float"
          data-class="status-tooltip"
          data-delay-hide="100"
          data-offset="{'bottom': -5, 'right': 5}" >
          {mainGlyph}
          <div className="badge">
              {statusCount}
          </div>
        </div>
        <ReactTooltip />
      </div>
    );
  } else {
    return (
      <div className="status-badge">
        {mainGlyph}
      </div>
    );
  }  
}
