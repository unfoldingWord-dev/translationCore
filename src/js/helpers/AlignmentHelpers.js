//helpers
import * as VerseObjectHelpers from './VerseObjectHelpers';
import * as ArrayHelpers from './ArrayHelpers';

/**
 * @description pivots alignments into bottomWords/targetLanguage verseObjectArray sorted by verseText
 * @param {Array} alignments - array of aligned word objects {bottomWords, topWords}
 * @param {Array} wordBank - array of topWords
 * @param {String} verseString - The string to base the bottomWords sorting
 * @param {Boolean} useVerseText - if true, then return parsed verse text if unaligned verse has changed, otherwise return null
 * @returns {Array} - sorted array of verseObjects to be used for verseText of targetLanguage
 */
export const merge = (alignments, wordBank, verseString, useVerseText=false) => {
  let verseObjects; // array to return
  // get the definitive list of verseObjects from the verse, unaligned but in order
  const unalignedOrdered = VerseObjectHelpers.getOrderedVerseObjectsFromString(verseString);
  // assign verseObjects with unaligned objects to be replaced with aligned ones
  verseObjects = JSON.parse(JSON.stringify(unalignedOrdered));
  //check each word in the verse string is also in the word bank or alignments
  const verseObjectsNotInAlignmentData = verseStringWordsContainedInAlignments(alignments, wordBank, verseObjects);
  if (verseObjectsNotInAlignmentData.length > 0) {
    if (hasAlignments(alignments)) { // if verse has some alignments
      const verseWordsJoined = verseObjectsNotInAlignmentData.map(({text}) => text).join(', ');
      throw {
        message: `The words "${verseWordsJoined}" from the target language verse are not in the alignment data.`,
        type: 'InvalidatedAlignments'
      };
    } else { // if verse had no alignments
      return useVerseText ? verseObjects : null; // use parsed verse text
    }
  }
  // each wordBank object should result in one verseObject
  for (let bottomWord of wordBank) {
    const verseObject = VerseObjectHelpers.wordVerseObjectFromBottomWord(bottomWord);
    const index = VerseObjectHelpers.indexOfVerseObject(unalignedOrdered, verseObject);
    if (index > -1) {
      verseObjects[index] = verseObject;
    }
    else {
      if (hasAlignments(alignments)) { // if verse has some alignments
        throw {message: `Word "${bottomWord.word}" is in wordBank, but missing from target language verse.`, type: 'InvalidatedAlignments'};
      } else { // if verse had no alignments
        return useVerseText ? verseObjects : null; // use parsed verse text
      }
    }
  }
  let indicesToDelete = [];
  // each alignment should result in one verseObject
  alignments.forEach(alignment => {
    const {topWords, bottomWords} = alignment;
    // each bottomWord results in a nested verseObject of tag: w, type: word
    // located inside innermost nested topWord/k verseObject
    let replacements = {};
    bottomWords.forEach(bottomWord => {
      const verseObject = VerseObjectHelpers.wordVerseObjectFromBottomWord(bottomWord);
      const index = VerseObjectHelpers.indexOfVerseObject(unalignedOrdered, verseObject);
      if (index === -1) {
        throw { message: "VerseObject not found in verseText while merging:" + JSON.stringify(verseObject), type: 'InvalidatedAlignments' };
      }
      replacements[index] = verseObject;
    });
    // each topWord results in a nested verseObject of tag: k, type: milestone
    const milestones = topWords.map(topWord =>
      VerseObjectHelpers.milestoneVerseObjectFromTopWord(topWord)
    );
    const indices = Object.keys(replacements);
    // group consecutive indexes so that they can be aggregated
    const groupedConsecutiveIndices = ArrayHelpers.groupConsecutiveNumbers(indices);
    // loop through groupedConsecutiveIndices to reduce and place where needed.
    groupedConsecutiveIndices.forEach(consecutiveIndices => {
      // map the consecutiveIndices to replacement verseObjects
      const replacementVerseObjects = consecutiveIndices.map(index => replacements[index]);
      // remove and use the first index in group to place the aligned verseObject milestone later
      const indexToReplace = consecutiveIndices.shift();
      // the rest of the consecutiveIndices need to be queued to be deleted later after shift
      indicesToDelete = indicesToDelete.concat(consecutiveIndices);
      // place the replacementVerseObjects in the last milestone as children
      milestones[milestones.length-1].children = replacementVerseObjects;
      // nest the milestones so that the first is the parent and each subsequent is nested
      const milestone = VerseObjectHelpers.nestMilestones(milestones);
      // replace the original verseObject from the verse text with the aligned milestone verseObject
      verseObjects[indexToReplace] = milestone;
    });
  });
  // deleteIndices that were queued due to consecutive bottomWords in alignments
  verseObjects = ArrayHelpers.deleteIndices(verseObjects, indicesToDelete);
  return verseObjects;
};

/**
 * check if there were any alignments
 * @param {Array} alignments
 * @return {boolean} true if an alignment was found
 */
export const hasAlignments = (alignments) => {
  const indexFirstAlignment = alignments.findIndex((alignment) => {
    return alignment.bottomWords.length > 0;
  });
  return indexFirstAlignment >= 0;
};

/**
 * Determines if the given verse objects from a string are contained in
 * the given alignments
 *
 * @param {Array} alignments - array of aligned word objects {bottomWords, topWords}
 * @param {Array} wordBank - array of unused topWords for aligning
 * @param {Object} verseObjects - verse objects from verse string to be checked
 */
export function verseStringWordsContainedInAlignments(alignments, wordBank, verseObjects) {
  return verseObjects.filter((verseObject) => {
    const checkIfWordMatches = function(verseObject) {
      return function({ word, occurrence, occurrences }) {
        const verseObjectWord = verseObject.text;
        const verseObjectOccurrence = verseObject.occurrence;
        const verseObjectOccurrences = verseObject.occurrences;
        return word === verseObjectWord &&
          occurrence === verseObjectOccurrence &&
          occurrences === verseObjectOccurrences;
      };
    };
    if (verseObject.type !== 'word') return false;
    const wordCheckerFunction = checkIfWordMatches(verseObject);
    const containedInWordBank = !!wordBank.find(wordCheckerFunction);
    const containedInAlignments = !!alignments.find(({ bottomWords }) => {
      return !!bottomWords.find(wordCheckerFunction);
    });
    return !containedInWordBank && !containedInAlignments;
  });
}

/**
 * @description pivots alignments into bottomWords/targetLanguage verseObjectArray sorted by verseText
 * @param {Array} verseObjects - array of aligned verseObjects [{milestone children={verseObject}}, ...]
 * @param {Array|Object|String} alignedVerse - optional verse to use for ordering alignments
 * @returns {Object} - object of alignments (array of alignments) and wordbank (array of unused words)
 */
export const unmerge = (verseObjects, alignedVerse) => {
  let baseMilestones = [], wordBank = [];
  let alignments = [];
  if (verseObjects && verseObjects.verseObjects) {
    verseObjects = verseObjects.verseObjects;
  }
  if (typeof alignedVerse !== 'string') {
    alignedVerse = VerseObjectHelpers.getWordList(alignedVerse);
  }
  for (let verseObject of verseObjects) {
    let alignment = getAlignmentForMilestone(baseMilestones, verseObject);
    if (!alignment) {
      alignment = {topWords: [], bottomWords: []};
      alignments.push(alignment);
      baseMilestones.push({alignment: alignment, milestone: verseObject});
    }
    addVerseObjectToAlignment(verseObject, alignment);
  }
  const alignmentUnOrdered = [];
  for (let _alignment of alignments) {
    if (_alignment.topWords.length > 0) {
      alignmentUnOrdered.push(_alignment);
    } else {
      wordBank = wordBank.concat(_alignment.bottomWords);
    }
  }
  let alignment = orderAlignments(alignedVerse, alignmentUnOrdered);
  return { alignment, wordBank};
};

/**
 * @description uses the alignedVerseString to order alignments
 * @param {String|Array} alignmentVerse - optional alignment verse
 * @param {Array} alignmentUnOrdered - alignments to order
 * @return {Array} ordered alignments if alignment string given, else unordered alignments
 */
export const orderAlignments = function (alignmentVerse, alignmentUnOrdered) {
  if (typeof alignmentVerse === 'string') {
    alignmentVerse = VerseObjectHelpers.getOrderedVerseObjectsFromString(alignmentVerse);
  } else {
    alignmentVerse = VerseObjectHelpers.getOrderedVerseObjects(alignmentVerse);
  }
  if (Array.isArray(alignmentVerse)) {
    let alignment = [];
    // order alignments
    for (let i = 0; i < alignmentVerse.length; i++) {
      const nextWord = alignmentVerse[i];
      let index = indexOfFirstMilestone(alignmentUnOrdered, nextWord);
      if ((index < 0) && (nextWord.type === 'word') && (i < alignmentVerse.length - 1)) {
        const wordAfter = alignmentVerse[i + 1];
        if (wordAfter.type === 'text') { // maybe this was punctuation split from word
          nextWord.text += wordAfter.text; // add possible punctuation
          index = indexOfFirstMilestone(alignmentUnOrdered, nextWord); // try again
        }
      }
      if (index >= 0) {
        alignment.push(alignmentUnOrdered[index]);
        alignmentUnOrdered.splice(index, 1); // remove item
      } else { // if not found, may be either an unaligned topWord or merged topWord
        if (nextWord.type === 'word') {
          index = indexOfMilestone(alignmentUnOrdered, nextWord);
          if (index < 0) { // if not found in unordered list, try already ordered
            index = indexOfMilestone(alignment, nextWord);
          }
          if (index < 0) { // if still not found in topWords, it's an unaligned topWord
            const wordObject = VerseObjectHelpers.alignmentObjectFromVerseObject(nextWord);
            alignment.push({topWords: [wordObject], bottomWords: []});
          }
        }
      }
    }
    if (alignmentUnOrdered.length > 0) {
      alignment = alignment.concat(alignmentUnOrdered);
    }
    return alignment;
  }
  return alignmentUnOrdered;
};

/**
 * compare occurrences of a and b, and handle conversion to int if necessary
 * @param a
 * @param b
 * @return {boolean}
 */
const compareOccurrences = function (a, b) {
  let sameOccurrence = (a.occurrence === b.occurrence);
  if (!sameOccurrence && a.occurrence && b.occurrence) {
    if (typeof a.occurrence !== typeof b.occurrence) { // one may be string and the other an int
      const occurrence1 = (typeof a.occurrence === 'string') ? parseInt(a.occurrence) : a.occurrence;
      const occurrence2 = (typeof b.occurrence === 'string') ? parseInt(b.occurrence) : b.occurrence;
      sameOccurrence = (occurrence1 === occurrence2) && (occurrence1 !== 0);
    }
  }
  return sameOccurrence;
};
/**
 * @description Returns index of the verseObject in the alignments first milestone (ignores occurrences since that can be off)
 * @param {Array} alignments - array of the alignments to search in
 * @param {Object} verseObject - verseObject to search for
 * @returns {Int} - the index of the verseObject
 */
export const indexOfFirstMilestone = (alignments, verseObject) => {
  let index = -1;
  if (verseObject.type === 'word') {
    index = alignments.findIndex(alignment => {
      if (alignment.topWords.length > 0) {
        const _verseObject = alignment.topWords[0];
        if (_verseObject.word === verseObject.text) {
          return compareOccurrences(_verseObject, verseObject);
        }
      }
      return false;
    });
  }
  return index;
};
/**
 * @description Returns index of the verseObject in the alignments milestone (ignores occurrences since that can be off)
 * @param {Array} alignments - array of the alignments to search in
 * @param {Object} verseObject - verseObject to search for
 * @returns {Int} - the index of the verseObject
 */
export const indexOfMilestone = (alignments, verseObject) => {
  let index = -1;
  if (verseObject.type === 'word') {
    index = alignments.findIndex(alignment => {
      for (let _verseObject of alignment.topWords) {
        if (_verseObject.word === verseObject.text) {
          return compareOccurrences(_verseObject, verseObject);
        }
      }
      return false;
    });
  }
  return index;
};

/**
 * @description find the alignment to use for this milestone.  If milestone has already been given an alignment, then
 *                use that one.  Otherwise return null.  This is needed because milestones are not always
 *                contiguous.
 * @param {Array} baseMilestones - already found base milestones.
 * @param {Object} newMilestone
 * @return {Object} previous Alignment if found - else null.
 */
const getAlignmentForMilestone = (baseMilestones, newMilestone) => {
  for (let baseMilestone of baseMilestones) {
    if (baseMilestone.alignment && sameMilestone(baseMilestone.milestone, newMilestone)) {
      return baseMilestone.alignment;
    }
  }
  return null;
};

/**
 * @description test to see if this is the same milestone (needed when milestones are not contiguous)
 * @param {Object} a
 * @param {Object} b
 * @return {boolean} true if same milestone
 */
const sameMilestone = (a, b) => {
  const same = (a.type === b.type) && (a.content === b.content) && (a.occurrence === b.occurrence);
  return same;
};

/**
 * @description adds verse object to alignment
 * @param {Object} verseObject
 * @param {Object} alignment
 */
export const addVerseObjectToAlignment = (verseObject, alignment) => {
  if (verseObject.type === 'milestone' && verseObject.children.length > 0) {
    const wordObject = VerseObjectHelpers.alignmentObjectFromVerseObject(verseObject);
    const duplicate = alignment.topWords.find(function (obj) {
      return (obj.word === wordObject.word) && (obj.occurrence === wordObject.occurrence) ;
    });
    if (!duplicate) {
      alignment.topWords.push(wordObject);
    }
    verseObject.children.forEach(_verseObject => {
      addVerseObjectToAlignment(_verseObject, alignment);
    });
  } else if (verseObject.type === 'word' && !verseObject.children) {
    const wordObject = VerseObjectHelpers.alignmentObjectFromVerseObject(verseObject);
    alignment.bottomWords.push(wordObject);
  }
};

