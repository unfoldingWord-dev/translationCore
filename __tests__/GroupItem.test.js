/* eslint-disable no-console */
/* eslint-env jest */
import React from 'react';
import GroupItem from '../src/js/components/groupMenu/GroupItem';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
// helpers
import * as statusBadgeHelpers from '../src/js/helpers/statusBadgeHelpers';

describe('Tests for GroupItem', () => {
  test('Test GroupItem', () => {
    const props = {
      bookName: 'Titus',
      selectionText: 'Paul',
      contextId: {
        reference: {
          chapterVerseMenu: true,
          text: 'Tit 1:1',
          verse: 'Paul'
        }
      },
      actions: {
          changeCurrentContextId: jest.fn()
      },
      statusBadge: statusBadgeHelpers.getStatusBadge(['ok', 'comment']),
      scrollIntoView: jest.fn(),
      inView: jest.fn(),
      active: true,
    };

    // when
    const wrapper = shallow(
      <GroupItem {...props} />
    );

    // then
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
