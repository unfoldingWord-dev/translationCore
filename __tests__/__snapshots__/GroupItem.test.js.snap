// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for GroupItem Test GroupItem 1`] = `
<div
  className="group-item active"
  onClick={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "var(--accent-color)",
      "borderBottom": "1px solid var(--background-color-dark)",
      "color": "var(--reverse-color)",
      "cursor": "pointer",
      "display": "flex",
      "height": 38,
      "padding": "10px 0",
      "zIndex": 1,
    }
  }
>
  <div
    className="status-badge-wrapper"
  >
    <div
      className="status-badge"
      data-class="status-tooltip"
      data-delay-hide="100"
      data-effect="float"
      data-html="true"
      data-offset="{'bottom': -5, 'right': 5}"
      data-place="bottom"
      data-tip="<span style=\\"color:var(--highlight-color);display:initial\\" class=\\"glyphicon glyphicon-comment\\" data-reactroot=\\"\\"></span>"
    >
      <Glyphicon
        bsClass="glyphicon"
        glyph="ok"
        key="ok"
        style={
          Object {
            "color": "var(--completed-color)",
            "display": "initial",
          }
        }
      />
      <div
        className="badge"
      >
        2
      </div>
    </div>
    <ReactTooltip
      insecure={true}
      resizeHide={true}
      wrapper="div"
    />
  </div>
  <span
    className="selection"
    data-class="selection-tooltip"
    data-delay-hide="100"
    data-effect="float"
    data-place="bottom"
    data-tip="Paul"
    data-type="dark"
  >
    <span
      style={
        Object {
          "display": "block",
          "overflow": "hidden",
          "padding": "0px 20px 0px 0px",
          "textOverflow": "ellipsis",
          "whiteSpace": "nowrap",
        }
      }
    >
      Tit 1:1 Paul
    </span>
  </span>
  <ReactTooltip
    insecure={true}
    resizeHide={true}
    wrapper="div"
  />
</div>
`;
